Subject:       Changes to boot from plain dm-crypt device (v3)
Patch Author:  Orson Teodoro <orsonteodoro@hotmail.com>
Patch Status:  Complete on 4.0.x
Date:          Mon Jul 19 01:02:36 PM PDT 2021 (Unix time: 1626724956)

v3 - reduce conditional, added USE_CRYPTSETUP=1
v2 - genkernel 4.0.10 compatibility
v1 - initial release

diff -urp genkernel-4.0.10.orig/defaults/initrd.scripts genkernel-4.0.10/defaults/initrd.scripts
--- genkernel-4.0.10.orig/defaults/initrd.scripts	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/defaults/initrd.scripts	2021-07-19 13:01:23.309965402 -0700
@@ -1706,7 +1706,7 @@ start_volumes() {
 			then
 				good_msg "ZFS pool ${ZFS_POOL} already imported."
 
-				if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_SWAP}" ]
+				if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_SWAP}" -o -n "${CRYPT_ROOT_PLAIN}" ]
 				then
 					good_msg "LUKS detected. Reimporting ${ZFS_POOL} ..."
 
@@ -1848,7 +1848,7 @@ openLUKS() {
 			KEYDEV_ERROR=0
 		else
 			LUKS_DEVICE=$(find_real_device "${LUKS_DEVICE}")
-			if [ -z "${LUKS_DEVICE}" ]
+			if [ -z "${LUKS_DEVICE}" -a -z "${CRYPT_ROOT_PLAIN}" ]
 			then
 				bad_msg "Looks like CRYPT_${TYPE} kernel cmdline argument is not set." ${CRYPT_SILENT}
 				DEV_ERROR=1
@@ -1856,7 +1856,7 @@ openLUKS() {
 			fi
 
 			setup_md_device ${LUKS_DEVICE}
-			if ! run cryptsetup isLuks ${LUKS_DEVICE}
+			if [ ! run cryptsetup isLuks ${LUKS_DEVICE} -a -z "${CRYPT_ROOT_PLAIN}" ]
 			then
 				bad_msg "The LUKS device ${LUKS_DEVICE} does not contain a LUKS header" ${CRYPT_SILENT}
 				DEV_ERROR=1
@@ -1951,7 +1951,12 @@ openLUKS() {
 					fi
 				fi
 				# At this point, keyfile or not, we're ready!
-				crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				if [ -n "${CRYPT_ROOT_PLAIN}" ]
+				then
+					crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} plainOpen ${CRYPT_ROOT_PLAIN} ${LUKS_NAME}"
+				else
+					crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				fi
 				crypt_filter_ret=$?
 
 				[ -e /dev/tty.org ] \
@@ -2314,7 +2319,7 @@ start_LUKS() {
 	[ -n "${CRYPT_ROOT_KEY}" ] && [ -z "${CRYPT_ROOT_KEYDEV}" ] \
 		&& sleep 6 && bootstrapKey "ROOT"
 
-	if [ -n "${CRYPT_ROOT}" ]
+	if [ -n "${CRYPT_ROOT}" -o -n "${CRYPT_ROOT_PLAIN}" ]
 	then
 		openLUKS "root"
 		if [ -n "${REAL_ROOT}" ]
diff -urp genkernel-4.0.10.orig/defaults/linuxrc genkernel-4.0.10/defaults/linuxrc
--- genkernel-4.0.10.orig/defaults/linuxrc	2020-07-20 09:05:15.000000000 -0700
+++ genkernel-4.0.10/defaults/linuxrc	2021-07-19 12:59:46.638990789 -0700
@@ -248,6 +248,10 @@ do
 			CRYPT_ROOT=${x#*=}
 			USE_CRYPTSETUP=1
 		;;
+		crypt_root_plain=*)
+			CRYPT_ROOT_PLAIN=${x#*=}
+			USE_CRYPTSETUP=1
+		;;
 		crypt_swap=*)
 			CRYPT_SWAP=${x#*=}
 			USE_CRYPTSETUP=1
Only in genkernel-4.0.10/defaults: linuxrc.orig
